runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -555,12 +557,140 @@ bool J9::ValuePropagation::transformUnsafeCopyMemoryCall(TR::Node *arraycopyNode
+   //
+   if (constraint->isClassObject() == TR_yes)
+      {
+      return TR_no;
+      }
+
+   // If the type is unresolved no information is available -
+   // VP has to assume that it might be a value type
+   //
+   TR::VPClassType *maybeUnresolvedType = constraint->getClassType();
+   if (maybeUnresolvedType == NULL)
+      {
+      return TR_maybe;
+      }
+
+   TR::VPResolvedClass *type = maybeUnresolvedType->asResolvedClass();
+   if (type == NULL)
+      {
+      return TR_maybe;
+      }
+
+   // Check for a type of java/lang/Object.  If the type is fixed to
+   // that class, it's not a value type; if it's not fixed, it could
+   // any subtype of java/lang/Object, which includes all value types
+   //
+   TR::Compilation *comp = TR::comp();
+   TR_OpaqueClassBlock *clazz = type->getClass();
 
+   if (clazz == comp->getObjectClassPointer())
+      {
+      return type->isFixedClass() ? TR_no : TR_maybe;
+      }
+
+   // Is the type either an abstract class or an interface (i.e., not a
+   // concrete class)?  If so, it might be a value type.
+   //
+   // Future refinements:
+   //   The JVM will have classes whose instances must be identity
+   //   objects implement the java/lang/IdentityObject interface, including
+   //   array classes.  Once that happens, the check for isClassArray can
+   //   be replaced with a test of whether the component class implements
+   //   IdentityObject, and the test of !isConcreteClass can be refined
+   //   to distinguish abstract classes that implement IdentityObject from
+   //   those that do not, as well as interfaces that extend IdentityObject.
+   //
+   if (!TR::Compiler->cls.isConcreteClass(comp, clazz))
+      {
+      return TR_maybe;
+      }
+
+   return TR::Compiler->cls.isValueTypeClass(clazz) ? TR_yes : TR_no;
    }
 
 void
 J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
    {
+   const bool isObjectEqualityCompare =
+      comp()->getSymRefTab()->isNonHelper(
+         node->getSymbolReference(),
+         TR::SymbolReferenceTable::objectEqualityComparisonSymbol);
+
+   if (isObjectEqualityCompare)
+      {
+      // Only constrain the call in the last run of vp to avoid handling the candidate twice if the call is inside a loop
+      if (lastTimeThrough())
+         {
+         addGlobalConstraint(node, TR::VPIntRange::create(this, 0, 1));
+         }
+      else
+         {
+         return;
+         }
+
+      bool lhsGlobal, rhsGlobal;
+      TR::Node *lhsNode = node->getChild(0);
+      TR::Node *rhsNode = node->getChild(1);
+      TR::VPConstraint *lhs = getConstraint(lhsNode, lhsGlobal);
+      TR::VPConstraint *rhs = getConstraint(rhsNode, rhsGlobal);
+
+      const TR_YesNoMaybe isLhsValue = isValue(lhs);
+      const TR_YesNoMaybe isRhsValue = isValue(rhs);
+      const bool areSameRef = getValueNumber(lhsNode) == getValueNumber(rhsNode)
+        || lhs != NULL && rhs != NULL && lhs->mustBeEqual(rhs, this);
+
```

Comment:
Some parentheses here might make the logic easier to read here

Reply:
Addressed this comment in commit cfd28b6087da2a2f87c89c904c1b7d5bc57c6567